design-documents/service-isolation/detailed-architecture.md

```diff
@@ -0,0 +1,87 @@
+# Detailed architecture of multi-service Magento
+
+## Overview
+
+The goal of this document is to gather requirements, understand data flow and propose architecture for multi-service Magento application.
+
+Please make sure to get familiar with [high-level service isolation vision](../service-isolation.md) before proceeding.
+
+![Gateway](img/GraphQL%20BFF.png)
+
+The client (e.g. browser) is interacting with the system using GraphQL. The only component of the system exposed publicly is Storefront Gateway, which is responsible for routing, full response caching and perimeter-level authorization.
+
+Storefront Gateway routes incoming queries between PWA, GraphQL Engine and Storefront Authentication Gateway.
+
+PWA's main responsibility is to [render React application on the server-side](../frontend/server-side-rendering.md). Resulting HTML page with React application becomes cacheable on the server.
+
+GraphQL server parses GraphQL query, enriches the request (explained later) and proxies calls to the Storefront services. It also manages Customer context JWT. GraphQL schema is defined here and is extensible, query resolution is based on generic resolvers and configuration for field mapping to Storefront APIs.
```

Comment:
For routing to Adobe/Magento SAAS API, lot of companies are even using GraphQL to export uniform interface and connect to other servers (routing). This will be based on use-case analysis, so we have two options

1. Either Frontend calls Storefront Gateway and Gateway provides Service Discovery to the client, and then client connect's directly to the SAAS API (specially if its Adobe/Magento hosted API). SAAS API can have different security requirements, we need to think about how we can handle this?

2. GraphQL exposes uniform interface, and routes the request to SAAS API; which means Frontend will be calling only GraphQL API. Although in this approach, we can manage the access/security in a uniform manner.

Reply:
1. This sounds reasonable to me. Will add as an option to the document. Are you okay with having this as an alternative option in the proposal?
2. Hiding Magento SAAS API behind GraphQL will require additional network hop. We can discuss if this is acceptable.

Reply:
yes, I believe both options should be considered and thought out, and be available on case by case basis as per my prior experience, there will not be one size fits all.

Reply:
Added section